File: v4/rust/iprange-livedb/src/cancellation.rs
//! Explicit cancellation shared with bounded operations.

use std::fmt;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use crate::error::{Error, Result};

/// Thread-safe cancellation token checked between bounded units of work.
#[derive(Clone)]
pub struct CancellationToken {
    cancelled: Arc<AtomicBool>,
    poll: Option<Arc<dyn Fn() -> bool + Send + Sync>>,
}

impl CancellationToken {
    /// Create an active token.
    pub fn new() -> Self {
        Self {
            cancelled: Arc::new(AtomicBool::new(false)),
            poll: None,
        }
    }

    /// Create a token that polls one binding-owned callback at every checkpoint.
    #[doc(hidden)]
    pub fn from_poll(poll: Arc<dyn Fn() -> bool + Send + Sync>) -> Self {
        Self {
            cancelled: Arc::new(AtomicBool::new(false)),
            poll: Some(poll),
        }
    }

    /// Request cancellation. Repeated requests are harmless.
    pub fn cancel(&self) {
        self.cancelled.store(true, Ordering::Release);
    }

    /// Report whether cancellation was requested.
    pub fn is_cancelled(&self) -> bool {
        self.cancelled.load(Ordering::Acquire) || self.poll.as_ref().is_some_and(|poll| poll())
    }

    pub(crate) fn check(&self) -> Result<()> {
        if self.is_cancelled() {
            Err(Error::Cancelled)
        } else {
            Ok(())
        }
    }

    pub(crate) fn requires_external_poll(&self) -> bool {
        self.poll.is_some()
    }
}

impl fmt::Debug for CancellationToken {
    fn fmt(&self, output: &mut fmt::Formatter<'_>) -> fmt::Result {
        output
            .debug_struct("CancellationToken")
            .field("cancelled", &self.cancelled.load(Ordering::Acquire))
            .field("polling", &self.poll.is_some())
            .finish()
    }
}

impl Default for CancellationToken {
    fn default() -> Self {
        Self::new()
    }
}

File: v4/go/cancellation.go
// Cancellation shared with bounded logical operations (Rust
// cancellation.rs parity). Checkpoints between bounded units of work
// report ErrorCancelled once Cancel has been requested; repeated Cancel
// calls are harmless.

package iprangedb

import "sync/atomic"

// CancellationToken is one shared cancellation flag. A nil token means
// no cancellation source: operations run uncancellable.
type CancellationToken struct {
	cancelled atomic.Bool
}

// NewCancellationToken returns an active token.
func NewCancellationToken() *CancellationToken {
	return &CancellationToken{}
}

// Cancel requests cancellation. Repeated requests are harmless.
func (t *CancellationToken) Cancel() {
	t.cancelled.Store(true)
}

// IsCancelled reports whether cancellation was requested.
func (t *CancellationToken) IsCancelled() bool {
	return t.cancelled.Load()
}

// check reports ErrorCancelled when cancellation was requested; it is the
// bounded-operation checkpoint hook. A nil token is the uncancellable
// checkpoint: the method value passes a nil receiver and reports nothing.
func (t *CancellationToken) check() error {
	if t == nil {
		return nil
	}
	if t.IsCancelled() {
		return &Error{Code: ErrorCancelled, Detail: "operation was cancelled"}
	}
	return nil
}

// hook returns the checkpoint hook of t, or nil for a nil token. A nil
// token must yield a nil hook: a method value over a nil receiver is
// itself non-nil, and worker sessions treat any non-nil hook as an
// external cancellation poll (Rust CancellationToken::
// requires_external_poll), turning every validation checkpoint into a
// parent round trip for an uncancellable session. Rust's fresh token
// has no poll hook and disables external polling; this accessor keeps
// the Go nil token equivalent.
func (t *CancellationToken) hook() func() error {
	if t == nil {
		return nil
	}
	return t.check
}
